File: Python/_0541_Reverse_String_II/Solution.py
# coding=utf-8
# Date: 2021/8/20 10:06


# 执行用时：32 ms, 在所有 Python3 提交中击败了88.25%的用户
# 内存消耗：14.9 MB, 在所有 Python3 提交中击败了92.02%的用户
class Solution:
    def reverseStr(self, s: str, k: int) -> str:
        res = ""
        for i in range(0, len(s), k * 2):
            res += s[i:min(i + k, len(s))][::-1]
            if i + k < len(s):
                res += s[i + k:min(i + k * 2, len(s))]
        return res


if __name__ == "__main__":
    print(Solution().reverseStr("abcd", 2))  # bacd
    print(Solution().reverseStr("abcd", 3))  # cbad
    print(Solution().reverseStr("abcd", 4))  # dcba
    print(Solution().reverseStr("abcd", 5))  # dcba

File: Java/src/_0541_Reverse_String_II/Solution.java
package _0541_Reverse_String_II;

/**
 * Date: 2021/8/20 9:58
 * 执行用时：1 ms, 在所有 Java 提交中击败了93.16%的用户
 * 内存消耗：38.6 MB, 在所有 Java 提交中击败了22.48%的用户
 */
public class Solution {

    public String reverseStr(String s, int k) {
        StringBuilder res = new StringBuilder();

        for (int i = 0; i < s.length(); i += 2 * k) {
            res.append(new StringBuilder(s.substring(i, Math.min(i + k, s.length()))).reverse());
            if (i + k < s.length()) res.append(s, i + k, Math.min(i + 2 * k, s.length()));
        }
        return res.toString();
    }

    public static void main(String[] args) {

        System.out.println((new Solution()).reverseStr("abcd", 2));  // bacd
        System.out.println((new Solution()).reverseStr("abcd", 3));  // cbad
        System.out.println((new Solution()).reverseStr("abcd", 4));  // dcba
        System.out.println((new Solution()).reverseStr("abcd", 5));  // dcba
    }
}
